app/views/application/_notebook.slim

```diff
@@ -13,6 +13,31 @@ javascript:
   -cells = jn["cells"].paginate(page: @page, per_page: 20)
 -else
   -cells = jn["cells"]
+
+// Calculate what each page's code cell id and health should start at
+ruby:
+  page_cell_starts = [0]
+  if (cells.respond_to? :total_pages)
+    jn["cells"].each_with_index do |cell, index|
+      if index % 20 == 0 && index != 0
+        if cell["cell_type"] == "code" && !cell["source"].blank? && defined?(code_cells) && code_cells
+          code_cell_number += 1
+          page_cell_starts.push(code_cell_number)
+        else
+          page_cell_starts.push(code_cell_number)
+        end
+      elsif cell["cell_type"] == "code" && !cell["source"].blank? && defined?(code_cells) && code_cells
+        code_cell_number += 1
+      end
+    end
+  end
+  code_cell_number = 0
+
+// Designate starting code_cell_number based on page parameter
+-if (cells.respond_to? :total_pages) && params[:page] && params[:page].present?
+  -code_cell_number = page_cell_starts[params[:page].to_i - 1]
```

Comment:
Why calculate the start index for every page and then lookup the appropriate cell number?  Why not just calculate the appropriate one, short circuit the loop and short circuit it so it stops counting so you aren't looping over the entire notebook once and then 20 rows of it.  This is just like, counting, but with extra steps ;)

Reply:
which, I should ad, means this code shouldn't run 99% of the time, because 99% of the time people will only go to the first page and click Run, so why do all of this work?